src/util/alloc/test_alloc.c

```diff
@@ -0,0 +1,214 @@
+
+    int f;
+    if(      j==0UL                                     ) f = 0;
+    else if( j==OUTSTANDING_MAX || (i+j)==2UL*alloc_cnt ) f = 1;
+    else                                                  f = (int)(fd_rng_uint( rng ) & 1UL);
+
+    if( !f ) { /* Malloc */
+
+      /* Pick the size and alignment randomly */
+
+      int   lg_align = fd_rng_int_roll( rng, lg_align_max+2 );
+      ulong align    = fd_ulong_if( lg_align==lg_align_max+1, 0UL, 1UL<<lg_align );
+
+      sz[j] = fd_rng_ulong_roll( rng, sz_max+1UL );
+      
+      /* Allocate it */
+
+      mem[j] = (uchar *)fd_alloc_malloc( alloc, align, sz[j] );
+
+      /* Check if the value is sane */
+
+      if( !sz[j] && mem[j] )
+        FD_LOG_ERR(( "On tile %lu, alloc(%lu,%lu) failed, got %lx, expected NULL", tile_idx, align, sz[j], (ulong)mem[j] ));
+
+      if( sz[j] && !mem[j] )
+        FD_LOG_ERR(( "On tile %lu, alloc(%lu,%lu) failed, got %lx, expected non-NULL", tile_idx, align, sz[j], (ulong)mem[j] ));
+
+      if( !align ) align = FD_ALLOC_MALLOC_ALIGN_DEFAULT;
+      if( !fd_ulong_is_aligned( (ulong)mem[j], align ) )
+        FD_LOG_ERR(( "On tile %lu, alloc(%lu,%lu) failed, got %lx (misaligned)", tile_idx, align, sz[j], (ulong)mem[j] ));
+
+      /* Fill it with a bit pattern unique to this allocation */
+
+      pat[j] = (tile_idx<<32) | i;
+      ulong b;
+      for( b=0UL; (b+7UL)<sz[j]; b+=8UL ) *((ulong *)(mem[j]+b)) = pat[j];
+      for( ; b<sz[j]; b++ ) mem[j][b] = ((uchar)tile_idx);
+
+      /* This allocation is now outstanding */
+
+      j++;
+
+    } else { /* Free */
+
+      /* Determine which outstanding allocation to free (note j>0 here) */
+
+      ulong k = fd_rng_ulong_roll( rng, j );
+
+      /* Validate the bit pattern was preserved between alloc and free */
+
+      ulong b;
+      for( b=0UL; (b+7UL)<sz[k]; b+=8UL )
+        if( (*(ulong *)(mem[k]+b))!=pat[k] ) FD_LOG_ERR(( "On tile %lu, memory corruption detected", tile_idx ));
+      for( ; b<sz[k]; b++ ) if( mem[k][b]!=((uchar)tile_idx) ) FD_LOG_ERR(( "On tile %lu, memory corruption detected", tile_idx ));
+
+      /* Free the allocation */
+
+      fd_alloc_free( alloc, mem[k] );
+
+      /* Remove from outstanding allocations */
+
+      j--;
+      sz [k] = sz [j];
+      mem[k] = mem[j];
+      pat[k] = pat[j];
+
+    }
+  }
+
+  fd_alloc_leave( alloc );
+  fd_rng_delete( fd_rng_leave( rng ) );
+  return 0;
+}
+
+int
+main( int     argc,
+      char ** argv ) {
+  fd_boot( &argc, &argv );
+
+  char const * name      = fd_env_strip_cmdline_cstr ( &argc, &argv, "--wksp",      NULL,      NULL );
+  ulong        alloc_cnt = fd_env_strip_cmdline_ulong( &argc, &argv, "--alloc-cnt", NULL, 1048576UL );
+  ulong        align_max = fd_env_strip_cmdline_ulong( &argc, &argv, "--align-max", NULL,     256UL );
+  ulong        sz_max    = fd_env_strip_cmdline_ulong( &argc, &argv, "--sz-max",    NULL,   73728UL );
+  ulong        tile_cnt  = fd_tile_cnt();
+
```

Comment:
if (1U == tile_cnt)  FD_LOG_NOTICE(("tiles not configured.. go read...")) ?

Reply:
Not necessary ... works fine single threaded (extensive single threaded testing with tile_cnt==1 was already done, as was tested with that and with cores on multi-socket high core count boxes, as was deliberately error injection into every code path the implementation to make sure the tester did in fact find atomic bugs in all code paths).